src/backend/common/DefaultMemoryManager.hpp

```diff
@@ -0,0 +1,139 @@
+
+        memory_info(memory_info &other)  = delete;
+        memory_info(memory_info &&other) = default;
+        memory_info &operator=(memory_info &other) = delete;
+        memory_info &operator=(memory_info &&other) = default;
+    };
+
+    memory_info &getCurrentMemoryInfo();
+
+   public:
+    DefaultMemoryManager(int num_devices, unsigned max_buffers, bool debug);
+
+    // Initializes the memory manager
+    virtual void initialize() override;
+
+    // Shuts down the memory manager
+    virtual void shutdown() override;
+
+    // Intended to be used with OpenCL backend, where
+    // users are allowed to add external devices(context, device pair)
+    // to the list of devices automatically detected by the library
+    void addMemoryManagement(int device) override;
+
+    // Intended to be used with OpenCL backend, where
+    // users are allowed to add external devices(context, device pair)
+    // to the list of devices automatically detected by the library
+    void removeMemoryManagement(int device) override;
+
+    void setMaxMemorySize();
+
+    /// Returns a pointer of size at least long
+    ///
+    /// This funciton will return a memory location of at least \p size
+    /// bytes. If there is already a free buffer available, it will use
+    /// that buffer. Otherwise, it will allocate a new buffer using the
+    /// nativeAlloc function.
+    void* alloc(bool user_lock, const unsigned ndims, dim_t *dims,
+                const unsigned element_size) override;
+
+    /// returns the size of the buffer at the pointer allocated by the memory
+    /// manager.
+    size_t allocated(void *ptr) override;
+
+    /// Frees or marks the pointer for deletion during the nex garbage
+    /// collection event
+    void unlock(void *ptr, bool user_unlock) override;
+
+    /// Frees all buffers which are not locked by the user or not being
+    /// used.
+    void signalMemoryCleanup() override;
+
+    void printInfo(const char *msg, const int device) override;
+    void usageInfo(size_t *alloc_bytes, size_t *alloc_buffers,
+                   size_t *lock_bytes, size_t *lock_buffers) override;
+    void userLock(const void *ptr) override;
+    void userUnlock(const void *ptr) override;
+    bool isUserLocked(const void *ptr) override;
+    size_t getMemStepSize() override;
+    void setMemStepSize(size_t new_step_size) override;
+    float getMemoryPressure() override;
+    bool jitTreeExceedsMemoryPressure(size_t bytes) override;
+
+   protected:
+    DefaultMemoryManager()                                  = delete;
+    ~DefaultMemoryManager()                                 = default;
+    DefaultMemoryManager(const DefaultMemoryManager &other) = delete;
+    // DefaultMemoryManager(const DefaultMemoryManager &&other) = default;
+    DefaultMemoryManager &operator=(const DefaultMemoryManager &other) = delete;
+    // DefaultMemoryManager &operator=(const DefaultMemoryManager &&other) =
+    //     default;
```

Comment:
remove or uncomment the move ctor/assignment op? Although this should be movable.